python3

ayushnawal opened the pull request:
Tested

ayushnawal commented:
@srevinsaju it's an issue not related to changes made in this pull request, porting the activity doesn't produce it.

reproduced the same with master too.

will try to fix it.

srevinsaju commented:
@ayushnawal I have fixed the error for you https://github.com/ayushnawal/make-them-fall-activity/pull/1. Please review

ayushnawal commented:
Thanks @srevinsaju for the fix :grin: 
 
tested

@quozl please review

srevinsaju commented:
Tested on Sugar 0.116 VM and Sugar DE

quozl commented:
Thanks.  On review there are many calls to pygame.mixer.init, these should all be avoided, as it is a requirement of Sugargame v1.2 that all calls to pygame module initialisation is handled by passing the modules to PygameCanvas, so that Sugargame does the initialisation at the right time.

srevinsaju commented on cardiac.py:
```diff
@@ -67,9 +67,9 @@ def run(self, gameDisplay, info):
             sound = False
-            print 'error with sound', err
+            print('error with sound', err)
 
```
This should either be removed or be moved to main.game.make (if errors are there while you removed it) . @quozl meant that `pygame.mixer.init()` should not be initialized many times. Also sugargame initializes the pygame modules within itself. so it is not explicit need to initialize the mixer (pygame's audio management system). It can be better be removed. (see my ports on `IqActivity` and `falabracman-activity` on how to do this. Falabracman has a mixer example too.

srevinsaju commented on fear.py:
```diff
@@ -65,9 +65,9 @@ def run(self, gameDisplay, info):
-        except Exception, err:
+        except Exception as err:
             sound = False
-            print 'error with sound', err
```
this one also

srevinsaju commented on howtoplay.py:
```diff
@@ -29,9 +29,9 @@ def run(self, gameDisplay, info):
-        except Exception, err:
+        except Exception as err:
             sound = False
-            print 'error with sound', err
```
This one :)

srevinsaju commented on impossible.py:
```diff
@@ -77,9 +77,9 @@ def run(self, gameDisplay, info):
             sound = False
-            print 'error with sound', err
+            print('error with sound', err)
 
```
Here

srevinsaju commented on inferno.py:
```diff
@@ -78,9 +78,9 @@ def run(self, gameDisplay, info):
             sound = False
-            print 'error with sound', err
+            print('error with sound', err)
 
```
.

srevinsaju commented on main.py:
```diff
@@ -50,9 +50,9 @@ def make(self):
             self.sound = False
-            print 'error with sound', err
+            print('error with sound', err)
 
```
.

srevinsaju commented on normal.py:
```diff
@@ -60,9 +60,9 @@ def run(self, gameDisplay, info):
             sound = False
-            print 'error with sound', err
+            print('error with sound', err)
 
```
.

srevinsaju commented on scorescreen.py:
```diff
@@ -49,9 +49,9 @@ def run(self, gameDisplay, score, gamenumber):
-            print 'error with sound', err
+            print('error with sound', err)
 
         if os.path.getsize("score.pkl") == 0:
```
.

ayushnawal commented:
Removed `pygame.mixer.init`  as suggested.
Learned how sugargame initializes the pygame modules within itself.
no errors so didn't move them to `main.game.make` 
Thanks @srevinsaju :grin: 

please review

srevinsaju commented:
Great @ayushnawal, did a code review, will test it in a while

ayushnawal commented:
Thanks @srevinsaju , also fixed `TypeError: integer argument expected, got float` , have a look.

quozl commented:
Thanks.  Reviewed.

* [ ] there is a pygame.init call left in main.game.make, and the main loop function given to Sugargame does not match the semantics of the main loop in the design; see the test activity for how it should be laid out,
* [ ] while you have removed pygame.mixer.init from several places, you haven't added it to the PygameCanvas modules list.

I don't think you've understood the Sugargame design or API yet.  Have another go.

ayushnawal commented:
@quozl  sincere apologies.
 removed all instance of `pygame.init` and added them to modules list in PygameCanvas

 Tried to match the semantics of test-activity 
* stop pygame animation by setting `crashed = True` on use of GTK stop button,
* Return on pygame.QUIT instead of setting `crashed = True`

let me know if it requires any other changes.

srevinsaju commented:
@ayushnawal Tested on sugar de, changes seems to be good. No logs, errors or warnings

quozl commented:
Tested 69413dc.  It didn't start.

```
Traceback (most recent call last):
  File "/usr/share/sugar/activities/MakeThemFall.activity/main.py", line 100, in make
    open('score.pkl', 'w+')
PermissionError: [Errno 13] Permission denied: 'score.pkl'
```

ayushnawal commented:
Thanks, @quozl I didn't face this error but saw the traceback above.

Although I faced the same error somewhere in the past, I searched over StackOverflow:
Got this [answer](https://stackoverflow.com/questions/36434764/permissionerror-errno-13-permission-denied) helpful.

Probably it occurred on your environment, can you try running activity using root permissions?

Please let me know if the issue is due to my changes because I was not able to reproduce it, neither @srevinsaju reported about it.

srevinsaju commented:
@ayushnawal I was able to reproduce the error while installing it system wide on sugar

ayushnawal commented:
@srevinsaju running  `ls -la` shows that there is write permissions for `score.pkl`

`-rw-r--r--  1 ayush ayush    28 Feb  5 21:38 score.pkl`

> while installing it system wide on sugar

also can you please explain what you exactly meant by this? :sweat_smile:

srevinsaju commented:
creating a bundle and installing it

ayushnawal commented:
Got the same error with `master` too!

> Traceback (most recent call last):
  File "/usr/share/sugar/activities/MakeThemFall.activity/main.py", line 108, in make
    open('score.pkl','w+')
IOError: [Errno 13] Permission denied: 'score.pkl'

When we run the activity as a `root` user, it works well but running it as a normal user displays the above traceback in logs, I think the normal user is not having permissions to both read and write the file.

What do you think @quozl ?

quozl commented:
Happy to explain.  I thought it was obvious, but clearly it is not.

It is https://github.com/sugarlabs/make-them-fall-activity/issues/22 ... and is caused by not using an appropriate directory for writing the file.  We've seen it before several times with other activities that were only ever designed for the OLPC XO laptop, which installs activity bundles in `~/Activities`.  When using Fedora, Debian, or Ubuntu on other systems the convention is to install activities system wide so that other users of the same system can use them.

Running as root cannot be a solution.

Here's what I suggest you do;

* search all other activities for score.pkl so you can see how other fixes have been applied, and find the best fix,
* find out what the score file is used for,
* if the score file is not used, remove the code associated,
* if the score file is used, change the path to on that is writable.

srevinsaju commented:
@ayushnawal, to make it intuitive for you, the fix is to implement a `write_file` method within activity

quozl commented:
> @ayushnawal, to make it intuitive for you, the fix is to implement a `write_file` method within activity

I don't think so.  The intention of a game high score file is to record the high score separately to the always present journal object, so that the high score is not lost when a child clears the journal.

srevinsaju commented:
score.pkl can be written to a file in the write file method, and use the respective `file_path`

quozl commented:
> score.pkl can be written to a file in the write file method, and use the respective `file_path`

No, that's wrong, as I said above.  Stick Hero contains a correct fix.  

```python
score_path = os.path.join(get_activity_root(), 'data', 'score.pkl')
```

Affected activities are this one, Flappy Birds, and 2 Cars.

Also should probably use JSON rather than pickle, for future portability.

srevinsaju commented:
I literally forgot about the existence of `get_activity_root` .

I agree, activities that use pkl are less likely to be flatpakked

quozl commented:
> I literally forgot about the existence of `get_activity_root` .

:grin:

> I agree, activities that use pkl are less likely to be flatpakked

Not really, someone preparing a flatpak can choose to patch the problem instead of upstreaming it here.  I'd prefer it be fixed here, but it is difficult sometimes for people to get their changes into a form they are happy with.

srevinsaju commented:
> Not really, someone preparing a flatpak can choose to patch the problem instead of upstreaming it here

Yup, it's possible, but all prefer the easy way of reducing the patches and hardwork of converting them to JSON. I would prefer it to be fixed here too.

ayushnawal commented:
Fixed the error:sweat_smile: 
Tested.
Looking good.

quozl commented:
Tested.  It spins consuming CPU power.  The graphics aren't positioned in the centre of screen.  The game goes fast enough for a teenager, but not slow enough for an elementary school child.  There's no controls to slow it down.  I'm not surprised, as it doesn't look like the game has been tested or released since it was written for GCI 2015, and our criteria for accepting new activities during GCI is much lower because of time pressure.

All in all, the game needs testing by many others before it is ready to be released.

I've reviewed the Pygame initialisation again;
* [ ] the pygame module as a whole is listed in the call to PygameCanvas; why do you list that and also list submodules?  Why not initialise the submodules as we normally do in Sugargame?
* [ ] there's no exit test after Gtk.main_iteration ... 

I'm a bit worried ... why do you keep doing this?  I've told you in other pull requests about these things.  It is almost as if I'm talking to a different person in each pull request.  Are you the same person?

ayushnawal commented:
> Tested. It spins consuming CPU power. The graphics aren't positioned in the centre of screen. The game goes fast enough for a teenager, but not slow enough for an elementary school child. There's no controls to slow it down. I'm not surprised, as it doesn't look like the game has been tested or released since it was written for GCI 2015, and our criteria for accepting new activities during GCI is much lower because of time pressure.

Agreed. We can give user options for choosing levels between **easy, medium, hard** before starting the game and after the choosing the type(normal, impossible, inferno etc.). I would like to handle it separately.
This PR fixes some important issues: 
* permission error of score.pkl (activity was not starting due to this)
* fix TypeError traceback
* fix deprecated set_icon method call for setting audio icon
* port activity to latest sugargame and python3

> I'm a bit worried ... why do you keep doing this? I've told you in other pull requests about these things. It is almost as if I'm talking to a different person in each pull request. Are you the same person?

Yes :sweat_smile:, sincere apologies, actually this problem occurs because i just review the changes made by me, and how the activity gets affected due to it, this module initialization issue should have been tackled before when these activities were ported from sugargame v1.1 to v1.2.

 I will especially take care of this now, also I wasn't aware of not adding pygame as a module, i just **naively** replaced `pygame.init` by adding it to the module list, sorry for this.

Fixing it now, Thanks :sweat_smile:

ayushnawal commented:
@quozl review required on [this](https://github.com/sugarlabs/make-them-fall-activity/pull/21#issuecomment-586995412)

quozl commented on activity.py:
```diff
@@ -47,7 +47,7 @@ def build_toolbar(self):
         separator.show()
 
-        button = ToolButton('speaker-muted-100')
+        button = ToolButton(icon_name='speaker-muted-100')
```
What is this change for?  If it is needed, then we should also change sugargame/test/TestActivity.py for the same reason.

ayushnawal replied:
the change was made by @srevinsaju to fix this error. However, I reviewed it again and changed it as it was in `TestActivity.py`. working perfectly.

srevinsaju replied:
Yup, we can drop that argument, positional argument corresponds to the kwargs. I am not sure why I did that, but tested it now
@ayushnawal Can you edit my commit and revert the changes at that line only?

ayushnawal replied:
@srevinsaju I have already pushed a commit with other changes also as @quozl suggested. pr is complete now I guess :grin:

quozl commented on activity.py:
```diff
@@ -61,16 +61,21 @@ def build_toolbar(self):
 
+    def _stop_cb(self, button):
+        self.game.crashed = True
+
```
Two blank lines at end of file.

ayushnawal replied:
fixed

quozl commented on main.py:
```diff
@@ -46,20 +48,12 @@ class game:
 
     def make(self):
```
We would normally call this method `run`, as that would better fit the usage in terms of Sugargame.

ayushnawal replied:
done

quozl commented:
Thanks.  Reviewed.  There are nine different event loops each calling `Gtk.main_iteration`, and eight of them;
* [ ] don't have a way to detect the GTK stop button being pressed, or being activated by accelerator,
* [ ] call `pygame.quit` instead of returning,
* [ ] call `sys.exit` instead of allowing the Sugar toolkit to cleanup,
* [ ] call `pygame.event.get` in addition to `pygame.event.poll`, without processing all events, thus leading to the problem of events being ignored.

Ask yourself how you missed this.  It could be instructive.  :grin:

ayushnawal commented:
changes made as requested!

tedious, though :grimacing:

quozl commented:
Thanks.  Reviewed.  Tested.

https://github.com/sugarlabs/make-them-fall-activity/issues/23 created for activity review comments above.

Yes, tedious.  It looks like each mode was a copy and paste, instead of creating a function, class, or file to contain the common code.  A redesign of modularity would be welcome.
